olive/model/handler/onnx.py

```diff
@@ -74,29 +74,58 @@ def prepare_session(
-        elif isinstance(execution_providers, str):
+        # eps given by arguments > user provided ep list > default eps
+        # Note by myguo: in DLIS scenarios, the inferenece_settings is something like:
+        #  {'execution_provider': [('MIGraphXExecutionProvider', {})]}
+        # and the provider_lists specified in perf_tuning is: ['ROCMExecutionProvider', 'MIGraphXExecutionProvider'].
+        # Originally, the inference_settings priority is higher than the execution_providers specified in arguments.
+        # as the result, when calling OnnxEvaluator.evaluate with the provider_list in baseline evaluation,
+        # the execution provider will be used as ["MIGraphXExecutionProvider"].
+        # later, the check_ort_fallback(session, execution_providers) will check the
+        # whether the underlying session's providers(MIGraphXExecutionProvider, CPUExecutionProvider) contains the
+        # execution_providers(ROCMExecutionProvider, MIGraphXExecutionProvider). In this case, the
+        # ROCMExecutionProvider is excluded when creating inference session. and the OliveEvaluationError exception
+        # will be raised.
+        # To fix this issue, we need to make sure the execution_providers specified in arguments has higher priority
+        if execution_providers:
+            provider_options = None
+        else:
+            execution_providers = inference_settings.get("execution_provider")
+            provider_options = inference_settings.get("provider_options")
+            if not execution_providers:
+                execution_providers = self.get_default_execution_providers(device)
+                provider_options = None
+
+        if isinstance(execution_providers, (str, tuple)):
             execution_providers = [execution_providers]
         else:
             # the execution_providers is a list
-            pass
-        inference_settings["execution_provider"] = execution_providers
+            assert isinstance(
+                execution_providers, list
+            ), f"execution_providers should be a list, got {execution_providers}"
 
-        if (device == Device.GPU) and (rank is not None) and not inference_settings.get("provider_options"):
-            inference_settings["provider_options"] = [
-                {"device_id": str(rank)} if ep == "CUDAExecutionProvider" else {} for ep in execution_providers
-            ]
+        # split the execution_providers and provider_options
+        execution_providers, provider_options = check_and_normalize_provider_args(
+            execution_providers, provider_options, ort.get_available_providers()
+        )
 
-        return get_ort_inference_session(self.model_path, inference_settings, self.use_ort_extensions)
+        if (device == Device.GPU) and (rank is not None):
+            for i, ep in enumerate(execution_providers):
+                if ep == "CUDAExecutionProvider" and not provider_options[i]:
```

Comment:
Wouldn't this condition return false if provider options is `{}`? Don't we want to provide the `device_id` even when it is `{}`?

Reply:
Actually, never mind. I misread the condition. This is true only when provider options is `{}`. So, we want to insert the device id if the user hasn't specified the provider options.

Reply:
maybe we could instead check for `provider_options.get(device_id) is not None` since for distributed models, we want to have device id always?

Reply:
This is the original behavior
https://github.com/microsoft/Olive/blob/main/olive/model/handler/onnx.py#L94-L97